python/ray/serve/router.py

```diff
@@ -41,382 +43,190 @@ class Query:
+
+        removed = current_replica_set - updated_replica_set
+        for removed_replica_handle in removed:
+            # NOTE(simon): Do we warn if there are still inflight queries?
+            # The current approach is no because the queries objectrefs are
+            # just used to perform backpressure. Caller should decide what to
+            # do with the object refs.
+            del self.in_flight_queries[removed_replica_handle]
+
+        # State changed, reset the round robin iterator
+        if len(added) > 0 or len(removed) > 0:
+            self.replica_iterator = itertools.cycle(
+                self.in_flight_queries.keys())
+            self.config_updated_event.set()
+
+    def _try_assign_replica(self, query: Query) -> Optional[ray.ObjectRef]:
+        for _ in range(len(self.in_flight_queries.keys())):
+            replica = next(self.replica_iterator)
+            if len(self.in_flight_queries[replica]
+                   ) >= self.max_concurrent_queries:
+                # This replica is overloaded, try next one
+                continue
+            ref = replica.handle_request.remote(query)
+            self.in_flight_queries[replica].add(ref)
+            return ref
+        return None
+
+    @property
+    def _all_query_refs(self):
+        return list(
+            itertools.chain.from_iterable(self.in_flight_queries.values()))
+
+    def _drain_completed_object_refs(self) -> int:
+        refs = self._all_query_refs
+        done, _ = ray.wait(refs, num_returns=len(refs), timeout=0)
+        for replica_in_flight_queries in self.in_flight_queries.values():
+            replica_in_flight_queries.difference_update(done)
+        return len(done)
+
+    async def assign_replica(self, query: Query) -> ray.ObjectRef:
+        """Given a query, submit it to a replica and returns the object ref.
+
+        This method will keep track of the in flight queries for each replicas
+        and only send a query to available replicas (determined by the backend
+        max_concurrent_quries value.)
+        """
+        # Happy case: there is a free replica.
+        assigned_ref = self._try_assign_replica(query)
+        if assigned_ref is not None:
+            return assigned_ref
+
+        # Maybe there exist a free replica, we just need to refresh our queries
+        # tracker.
+        num_finished = self._drain_completed_object_refs()
+        # All replicas are really busy, let's wait for one query to complete or
+        # the config is updated.
+        if num_finished == 0:
+            await asyncio.wait(
+                self._all_query_refs + [self.config_updated_event.wait()],
+                return_when=asyncio.FIRST_COMPLETED)
+            if self.config_updated_event.is_set():
+                self.config_updated_event.clear()
+        # We are pretty sure a free replica is ready now, let's recurse and
+        # assign this query a replica.
+        return await self.assign_replica(query)
```

Comment:
I'd prefer not to use recursion here to avoid any issues with Python stack limits in case of weird behavior. The logic is potentially easier to follow that way too.